File: packages/llm-test-harness/src/metrics/json-path.ts
import type { MetricFn, MetricScore } from '../types.js'

/**
 * Resolves a dot-notation path (e.g. "user.address.city") in a nested object.
 * Array indices are supported: "items.0.name"
 */
function resolvePath(obj: unknown, path: string): { found: true; value: unknown } | { found: false } {
  const parts = path.split('.')
  let current: unknown = obj
  for (const part of parts) {
    if (current === null || current === undefined) return { found: false }
    if (typeof current !== 'object' && !Array.isArray(current)) return { found: false }
    current = (current as Record<string, unknown>)[part]
    if (current === undefined) return { found: false }
  }
  return { found: true, value: current }
}

/**
 * Asserts that parsed JSON at `path` equals `expected`.
 *
 * @example
 * // Response: '{"user": {"name": "Alice", "age": 30}}'
 * Metrics.jsonPath('user.name', 'Alice')
 * Metrics.jsonPath('user.age', 30)
 */
export class JsonPath implements MetricFn {
  readonly name: string
  private path: string
  private expected: unknown

  constructor(path: string, expected: unknown) {
    this.path = path
    this.expected = expected
    this.name = `JsonPath(${path})`
  }

  evaluate(text: string): MetricScore {
    let parsed: unknown
    try {
      parsed = JSON.parse(text)
    } catch {
      return {
        name: this.name,
        pass: false,
        score: 0,
        reason: 'Input is not valid JSON',
      }
    }

    const result = resolvePath(parsed, this.path)
    if (!result.found) {
      return {
        name: this.name,
        pass: false,
        score: 0,
        reason: `Path "${this.path}" not found in JSON`,
      }
    }

    const actual = result.value
    const pass = JSON.stringify(actual) === JSON.stringify(this.expected)
    return {
      name: this.name,
      pass,
      score: pass ? 1 : 0,
      reason: pass ? undefined : `Expected ${JSON.stringify(this.expected)}, got ${JSON.stringify(actual)}`,
    }
  }
}

File: packages/llm-test-harness-python/llm_test_harness/metrics/json_path.py
from __future__ import annotations

import json
from typing import Any

from ..types import MetricFn, MetricScore


def _resolve_path(obj: Any, path: str) -> tuple[bool, Any]:
    parts = path.split(".")
    current = obj
    for part in parts:
        if current is None:
            return False, None
        if isinstance(current, dict):
            if part not in current:
                return False, None
            current = current[part]
        elif isinstance(current, list):
            try:
                current = current[int(part)]
            except (ValueError, IndexError):
                return False, None
        else:
            return False, None
    return True, current


class JsonPath(MetricFn):
    """Assert a value at a dot-notation path in the parsed JSON response.

    Example::

        # Response: '{"user": {"name": "Alice", "age": 30}}'
        Metrics.json_path("user.name", "Alice")
        Metrics.json_path("user.age", 30)
        Metrics.json_path("items.0.id", 42)
    """

    def __init__(self, path: str, expected: Any) -> None:
        self._path = path
        self._expected = expected

    @property
    def name(self) -> str:
        return f"JsonPath({self._path})"

    def evaluate(self, text: str) -> MetricScore:
        try:
            parsed = json.loads(text)
        except json.JSONDecodeError:
            return MetricScore(name=self.name, passed=False, score=0.0, reason="Input is not valid JSON")

        found, actual = _resolve_path(parsed, self._path)
        if not found:
            return MetricScore(name=self.name, passed=False, score=0.0, reason=f'Path "{self._path}" not found in JSON')

        passed = json.dumps(actual, sort_keys=True) == json.dumps(self._expected, sort_keys=True)
        reason = None if passed else f"Expected {json.dumps(self._expected)}, got {json.dumps(actual)}"
        return MetricScore(name=self.name, passed=passed, score=1.0 if passed else 0.0, reason=reason)
